src/virtio-pci.c

```diff
@@ -0,0 +1,266 @@
+    else if (select == 1)
+        dev->guest_feature |= (uint64_t) feature << 32;
+}
+
+static void virtio_pci_reset(struct virtio_pci_dev *dev)
+{
+    /* TODO: virtio pci reset */
+}
+
+static void virtio_pci_write_status(struct virtio_pci_dev *dev)
+{
+    uint8_t status = dev->config.common_cfg.device_status;
+    if (status == 0) {
+        virtio_pci_reset(dev);
+    }
+}
+
+static void virtio_pci_select_virtq(struct virtio_pci_dev *dev)
+{
+    uint16_t select = dev->config.common_cfg.queue_select;
+    struct virtio_pci_common_cfg *config = &dev->config.common_cfg;
+
+    if (select < config->num_queues) {
+        uint64_t offset = offsetof(struct virtio_pci_common_cfg, queue_size);
+        memcpy((void *) config + offset, &dev->vq[select].info,
+               sizeof(struct virtq_info));
+    } else {
+        config->queue_size = 0;
+    }
+}
+
+static void virtio_pci_enable_virtq(struct virtio_pci_dev *dev)
+{
+    uint16_t select = dev->config.common_cfg.queue_select;
+    virtq_enable(&dev->vq[select]);
+}
+
+static void virtio_pci_disable_virtq(struct virtio_pci_dev *dev)
+{
+    uint16_t select = dev->config.common_cfg.queue_select;
+    virtq_disable(&dev->vq[select]);
+}
+
+static void virtio_pci_space_write(struct virtio_pci_dev *dev,
+                                   void *data,
+                                   uint64_t offset,
+                                   uint8_t size)
+{
+    if (offset < offsetof(struct virtio_pci_config, dev_cfg)) {
+        memcpy((void *) &dev->config + offset, data, size);
+        switch (offset) {
+        case VIRTIO_PCI_COMMON_DFSELECT:
+            virtio_pci_select_device_feature(dev);
+            break;
+        case VIRTIO_PCI_COMMON_GFSELECT:
+            virtio_pci_write_guest_feature(dev);
+            break;
+        case VIRTIO_PCI_COMMON_STATUS:
+            virtio_pci_write_status(dev);
+            break;
+        case VIRTIO_PCI_COMMON_Q_SELECT:
+            virtio_pci_select_virtq(dev);
+            break;
+        case VIRTIO_PCI_COMMON_Q_ENABLE:
+            if (dev->config.common_cfg.queue_enable)
+                virtio_pci_enable_virtq(dev);
+            else
+                virtio_pci_disable_virtq(dev);
+            break;
+        default:
+            if (offset >= VIRTIO_PCI_COMMON_Q_SIZE &&
+                offset <= VIRTIO_PCI_COMMON_Q_USEDHI) {
+                uint16_t select = dev->config.common_cfg.queue_select;
+                uint64_t info_offset = offset - VIRTIO_PCI_COMMON_Q_SIZE;
+                if (select < dev->config.common_cfg.num_queues)
+                    memcpy((void *) &dev->vq[select].info + info_offset, data,
+                           size);
+            }
+            /* guest notify buffer avail */
+            else if (offset ==
+                     offsetof(struct virtio_pci_config, notify_data)) {
+                virtq_handle_avail(&dev->vq[dev->config.notify_data.vqn]);
+            }
+            break;
+        }
+
```

Comment:
You can add a `return` statement here and then the followed `else` can be eliminated.